pkg/kube/krt/README.md

```diff
@@ -0,0 +1,157 @@
+
+Example computing a list of all container names across all pods:
+```go
+ContainerNames := krt.NewManyCollection[string](func(ctx krt.HandlerContext, pod *v1.Pod) (res []string) {
+    for _, c := range pod.Spec.Containers {
+      res = append(res, c.Name)
+    }
+    return res
+}) // Results in a Collection[string]
+```
+
+Example computing a list of service endpoints.
+```go
+Endpoints := krt.NewManyCollection[Endpoint](func(ctx krt.HandlerContext, svc *v1.Service) (res []Endpoint) {
+    for _, c := range krt.Fetch(ctx, Pods, krt.FilterLabel(svc.Spec.Selector)) {
+      res = append(res, Endpoint{Service: svc.Name, Pod: pod.Name, IP: pod.status.PodIP})
+    }
+    return res
+}) // Results in a Collection[Endpoint]
+```
+
+As a rule of thumb, if your `Collection` type is a list, you most likely should be using a different type to flatten the list.
+An exception to this would be if the list represents an atomic set of items that are never queried independently;
+in these cases, however, it is probably best to wrap it in a struct.
+For example, to represent the set of containers in a pod, we may make a `type PodContainers struct { Name string, Containers []string }` and have a
+`Collection[PodContainers]` rather than a `Collection[[]string]`.
+
+In theory, other forms could be expressed such as `func(input1 I1, input2 I2) *O`.
+However, there haven't yet been use cases for these more complex forms.
+
+### Transformation constraints
+
+In order for the framework to properly handle dependencies and events, transformation functions must adhere by a few properties.
+
+Basically, Transformations must be stateless and idempotent.
+* Any querying of other `Collection`s _must_ be done through `krt.Fetch`.
+* Querying other data stores that may change is not permitted.
+* Querying external state (e.g. making HTTP calls) is not permitted.
+* Transformations _may_ be called at any time, including many times for the same inputs. Transformation functions should not make any assumptions about calling patterns.
+
+### Fetch details
+
+In addition to simply fetching _all_ resources from a collection, a filter can be provided.
+This is more efficient than filtering outside of `Fetch`, as the framework can filter un-matched objects earlier, skipping redundant work.
+The following filters are provided
+
+* `FilterName(name, namespace)`: filters an object by Name and Namespace.
+* `FilterNamespace(namespace)`: filters an object by Namespace.
+* `FilterKey(key)`: filters an object by key.
+* `FilterLabel(labels)`: filters to only objects that match these labels.
+* `FilterSelects(labels)`: filters to only objects that **select** these labels. An empty selector matches everything.
+* `FilterSelectsNonEmpty(labels)`: filters to only objects that **select** these labels. An empty selector matches nothing.
+* `FilterGeneric(func(any) bool)`: filters by an arbitrary function.
```

Comment:
IIUC, these are applied on client side, not on kube-apiserver. 

Can this lib support this case: when an object that is fetched update label, the caller of the fetch function can be notified and re-run.

Reply:
Yes, they are client side. The library wraps around existing informers, so server-side filters can be applied to those.

The library does support this case